### PR TITLE
check code change

### PR DESCRIPTION
**Pull request by moshe-kabala**
- Add file checksup function
- Add CompereCode and push code for code push and project push
- use push code to check if code is change

### CODE REVIEW

File: pkg/code/utils.go [hunk 1]
```diff
@@ -143,6 +143,92 @@ func GetAndUpdateCodeIntegrationIfNotExists(ctx context.Context, workspaceConfig
...
+		}
+	}
+	codeIntegrationVersion, err := AddCodeIntegrationVersion(ctx, tarGzFile, codeIntegrationId, entryFile, secretId)
+	if err != nil {
+		return false, nil, err
+	}
+	return true, codeIntegrationVersion, nil
+}
+
+func CompereCodeVersion(ctx context.Context, compereVersion *CodeIntegrationVersion, tarGzFile *os.File, entryFile, secretId string) (bool, error) {
```
Comment (YuvalFeldman) [inline pkg/code/utils.go:179]:
```suggestion
func CompareCodeVersion(ctx context.Context, compareVersion *CodeIntegrationVersion, tarGzFile *os.File, entryFile, secretId string) (bool, error) {
```

File: pkg/code/utils.go [hunk 2]
```diff
@@ -143,6 +143,92 @@ func GetAndUpdateCodeIntegrationIfNotExists(ctx context.Context, workspaceConfig
...
+	return latestChecksum != newChecksum, nil
+}
+
+func IsCodeParseFailed(codeIntegrationVersion *CodeIntegrationVersion) bool {
+	return codeIntegrationVersion.TestStatus == tensorleapapi.TESTSTATUS_TEST_FAIL
+}
+
+func IsCodeParsing(codeIntegrationVersion *CodeIntegrationVersion) bool {
+	return codeIntegrationVersion.TestStatus == tensorleapapi.TESTSTATUS_DURING_TEST || codeIntegrationVersion.TestStatus == tensorleapapi.TESTSTATUS_BEFORE_TEST
 }
```
Comment (YuvalFeldman) [inline pkg/code/utils.go:234]: what about finished succesfully?

**Review by YuvalFeldman — APPROVED** [review]

File: cmd/code/push.go [hunk 3]
```diff
@@ -56,6 +57,9 @@ func NewPushCmd() *cobra.Command {
 					code.PrintCodeIntegrationVersionParserErr(codeIntegrationVersion)
 					return fmt.Errorf("code parsing failed")
 				}
+			} else if watch && code.IsCodeParseFailed(currentVersion) {
```
[the remark this answers is not in the document]
Comment (moshe-kabala) [reply cmd/code/push.go:60]: There is no else if watch = false
 if true I covered the all the cases (`code.IsCodeParseFailed(currentVersion)` is just  for readability)

File: pkg/code/utils.go [hunk 4]
```diff
@@ -143,6 +143,92 @@ func GetAndUpdateCodeIntegrationIfNotExists(ctx context.Context, workspaceConfig
...
+}
+
+func PushCode(ctx context.Context, force bool, codeIntegrationId string, tarGzFile *os.File, entryFile, secretId string) (pushed bool, current *CodeIntegrationVersion, err error) {
+	if !force {
+		log.Info("Checking if code changed")
+
+		latestVersion, err := GetLatestVersion(ctx, codeIntegrationId)
+
+		if err != nil {
+			log.Warnf("Failed to get latest code integration version: %v", err)
```
[the remark this answers is not in the document]
Comment (moshe-kabala) [reply pkg/code/utils.go:158]: No

File: pkg/code/utils.go (hunk 2 above)
Comment (moshe-kabala) [reply pkg/code/utils.go:234]: No need for now

**Pull request merged**